ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,380 @@
+
+  EpOr af1 af2 -> let
+    fields1 = flattenOrs af1
+    fields2 = flattenOrs af2
+
+    decls = mkTypesForField <$> (indexEmptyFields $ fields1 <> fields2)
+    typeOptions = (\(_, (b, _), _) ->  b) <$> decls
+      -- Throws away information about if the field is optional. Need to check if it is relavant in 'Or' structures
+    thisType = TsType typename (TsSum typeOptions)
+    thisDecls = (\(_, _, c) -> c) <$> decls
+    in thisType : (concat thisDecls)
+
+  EpList a -> let
+    elementTypeName = typename <> "_" <> "item"
+    elementTypes = mkTypeFor elementTypeName a
+    thisType = TsType typename (TsAlias $ TsList $ TsCustom elementTypeName)
+    in thisType : elementTypes
+
+  EpSet a -> let
+    elementTypeName = typename <> "_" <> "item"
+    elementTypes = mkTypeFor elementTypeName a
+    thisType = TsType typename (TsAlias $ TsList $ TsCustom elementTypeName)
+    in thisType : elementTypes
+
+  EpOption a -> let
+    elementTypeName = typename <> "_" <> "item"
+    elementTypes = mkTypeFor elementTypeName a
+    thisType = TsType typename (TsAlias $ TsOption $ TsCustom elementTypeName)
+    in thisType : elementTypes
+
+  EpMap k v -> let
+    (kTypename, kDecls) = let
+      kTypename_ = typename <> "_" <> "key"
+      in (kTypename_, mkTypeFor kTypename_ k)
+
+    (vTypename, vDecls) = let
+      vTypename_ = typename <> "_" <> "val"
+      in (vTypename_, mkTypeFor vTypename_ v)
+
+    thisType = TsType typename (TsAlias $ TsMap (TsCustom kTypename) (TsCustom vTypename))
+    in thisType : (kDecls <> vDecls)
+
+  EpBigMap k v -> let
+    (kTypename, kDecls) = let
+      kTypename_ = typename <> "_" <> "key"
+      in (kTypename_, mkTypeFor kTypename_ k)
+
+    (vTypename, vDecls) = let
+      vTypename_ = typename <> "_" <> "val"
+      in (vTypename_, mkTypeFor vTypename_ v)
+
+    thisType = TsType typename (TsAlias $ TsBigMap (TsCustom kTypename) (TsCustom vTypename))
+    in thisType : (kDecls <> vDecls)
+
+  EpString -> [TsType typename (TsAlias $ TsString)]
+  EpInt -> [TsType typename (TsAlias $ TsNumber)]
+  EpNat -> [TsType typename (TsAlias $ TsNumber)]
+  EpUnit -> [TsType typename (TsAlias $ TsUnit)]
+  EpSignature -> [TsType typename (TsAlias $ TsString)]
+  EpOperation -> [TsType typename (TsAlias $ TsString)]
+  EpContract _ -> [TsType typename (TsAlias $ TsString)]
+  EpLambda _ _ -> [TsType typename $ TsAlias TsString]
+  EpBytes -> [TsType typename $ TsAlias TsString]
+  EpAddress -> [TsType typename $ TsAlias TsString]
+  EpKey -> [TsType typename $ TsAlias TsString]
+  EpChainId -> [TsType typename $ TsAlias TsString]
+  EpMutez -> [TsType typename $ TsAlias TsNumber]
+  EpBool -> [TsType typename $ TsAlias TsBool]
+  EpKeyHash -> [TsType typename $ TsAlias TsString]
+  EpTimestamp -> [TsType typename $ TsAlias TsString]
+
+  where
+
+    hasAtleastOneNamedField :: [(Fieldname, EpType)] -> Bool
+    hasAtleastOneNamedField fs = isJust $ find ((> 0) . length . fst) fs
```

Comment:
`length` is O(n), so I'd avoid using it. Suggestion:
```haskell
hasAtLeastOneNamedField = isJust $ find (not . null . fst)
```
n.b.: Notice I changed `Atleast` with `AtLeast` as well.

Reply:
:facepalm: Fixed both.